### ISSUE TITLE
Don't repeat keyspace and table name for every column in the results (relevant for non prepared statements)

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#17788]
An idea due to @wprzytula.

In the CQL protocol (see, for example, https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec), when multiple column values are returned in a response, there is a "column_spec" for each one. This column_spec contains the column's name and type - but **may** be preceded by the keyspace and table name.  Since normally (as far as I know, actually **always**) all result columns come from the same table, there is also an option in the CQL protocol to enable a `GLOBAL_TABLES_SPEC` bit, and then you can write the keyspace and table name just once instead of for every column.

The transport/server.cc seems to have the code to print the output correctly if this flag is enabled, but I'm not sure it is. The only place I could find this flag being set is in the `prepared_metadata` constructor (cql/result_set.cc) which has:
```c++
    if (!names.empty() && column_specification::all_in_same_table(_names)) {
        _flags.set<flag::GLOBAL_TABLES_SPEC>();
    }
```
I don't see such code in `metadata`.

Is it possible that we use this optimization for prepared statements (`prepared_metadata`) but not unprepared statements (`metadata`)? Why? If the answer is yes, it's a good start (because the performance of prepared statements is most important), but why should unprepared statements waste more network bandwidth when repeating the ks/table name multiple times?

We should begin by testing whether in practice Scylla uses the global or per-column table name in responses - for both prepared and unprepared statements. Then we can consider if we need to use GLOBAL_TABLES_SPEC in more cases.

Maybe we'll discover that this is a non-issue. But it's possible that it is.

**Comment by wprzytula** [comment]
There is an opened PR to Rust driver https://github.com/scylladb/scylla-rust-driver/pull/956 that is built on an assumption that table specs for all columns in one response are always the same. While investigating the described issue, please verify this assumption as well.

**Comment by nyh** [comment]
Right. That's another interesting question - why do we even need to check `all_in_same_table(_names)` in the code, is there even a situation in Scylla where it's possible to return in one query columns from _different_ tables? I can't think of any situation where this is possible. Maybe the protocol was designed with the possibility of joins in mind, but the implementation doesn't have this possibility. If this is true, Scylla could always use `GLOBAL_TABLES_SPEC`, unconditionally. But we need to prove that this is true. It's probably true because as @wprzytula said above the Rust driver, and apparently also the Python driver, assume it to be true.

**Comment by piodul** [comment]
Just a note about prepared statements. After preparing a statement, the driver receives the column_specs for the statement and can cache it. Later, when executing the prepared statement, it can request the DB to skip sending column_specs in response. This is what most drivers (if not all) do AFAIK, so performance conscious users are not affected by this issue too heavily.

However, doing what the issue suggests will still improve efficiency for statement preparation and sending unprepared queries.

**Comment by mykaul** [comment]
Optimistically setting this to 6.1. Let me know if we need to defer it to backlog.

**Comment by nyh** [comment]
Comment from @mykaul that he accidentally put in https://github.com/scylladb/scylladb/issues/7397:

Example of a simple query:
```
SELECT * FROM system.local WHERE key='local'
```

It returns 1 row, 19 columns, each of them has 15 bytes of <length:6, 'system', length:5, 'local'> column specification. So we could have saved 18*15 bytes - 270 bytes. Not that much from a 7932 bytes message (in my example) - and compression, which wasn't used here, could have helped.
But I agree it sounds like an easy win.

A different example is a 'PREPARED' result that I got - 60 columns, each had this 22 bytes lengthed KS name and 15 bytes table name, so we could have saved (2 + 22 + 2 + 15) * 59 = 2419 bytes. The whole message (again uncompressed for some reason) was 3409 bytes.

**Comment by mykaul** [comment]
> Right. That's another interesting question - why do we even need to check `all_in_same_table(_names)` in the code, is there even a situation in Scylla where it's possible to return in one query columns from _different_ tables? I can't think of any situation where this is possible. Maybe the protocol was designed with the possibility of joins in mind, but the implementation doesn't have this possibility. If this is true, Scylla could always use `GLOBAL_TABLES_SPEC`, unconditionally. But we need to prove that this is true. It's probably true because as [@wprzytula](https://github.com/wprzytula) said above the Rust driver, and apparently also the Python driver, assume it to be true.

@nyh - I think batch (see https://github.com/scylladb/scylla-rust-driver/issues/1134 ) is a (single?) example where we can have multiple tables.

**Comment by mykaul** [comment]
Dug into this just a bit. Why does the GoCQL driver disables SkipMetadata in its queries (https://github.com/scylladb/gocql/blob/e12494df329053d63aa205ffd4d7082587ea72bd/conn.go#L1793 ) ?

Because *AWS Managed Cassandra* has hard time with these READ queries! see https://github.com/apache/cassandra-gocql-driver/issues/1389

This is just ridiculous. We should identify early on if it's Scylla and stop this nonsense.

**Comment by mykaul** [comment]
> Dug into this just a bit. Why does the GoCQL driver disables SkipMetadata in its queries (https://github.com/scylladb/gocql/blob/e12494df329053d63aa205ffd4d7082587ea72bd/conn.go#L1793 ) ?
> 
> Because _AWS Managed Cassandra_ has hard time with these READ queries! see [apache/cassandra-gocql-driver#1389](https://github.com/apache/cassandra-gocql-driver/issues/1389)
> 
> This is just ridiculous. We should identify early on if it's Scylla (via `isScyllaConn()` ?) and stop this nonsense.
> 
> This is not related to global table specs, but is an additional item for non prepared statements, especially system queries.

Sadly, we have https://github.com/scylladb/gocql/pull/275 which I don't believe we've fixed on the server side.

**Comment by wprzytula** [comment]
> Right. That's another interesting question - why do we even need to check `all_in_same_table(_names)` in the code, is there even a situation in Scylla where it's possible to return in one query columns from _different_ tables? I can't think of any situation where this is possible. Maybe the protocol was designed with the possibility of joins in mind, but the implementation doesn't have this possibility. If this is true, Scylla could always use `GLOBAL_TABLES_SPEC`, unconditionally. But we need to prove that this is true. It's probably true because as [@wprzytula](https://github.com/wprzytula) said above the Rust driver, and apparently also the Python driver, assume it to be true.

Not directly related to the case of ResultMetadata, but it shocked us in the Rust driver: https://github.com/scylladb/scylla-rust-driver/issues/1134.

Tl;dr: in PreparedMetadata, for a textually-constructed batch, bind markers can refer to different columns. As we assumed that TableSpec will always be the same for all columns - both for ResultMetadata and PreparedMetadata - we introduced an assertion of that, which showed to be a bug.

Just FYI, to have this in mind in case the checks in Core for global TableSpecs (e.g., `all_in_same_table(_names)` that you mentioned) are shared between ResultMetadata and PreparedMetadata.
